### ISSUE TITLE
ODS where fid=3 does not return data

### ISSUE THREAD

**Issue by sfischaber** [pramsey/pgsql-ogr-fdw#119]
Is it possible to use the fid column to filter a select statement?  I'm trying to get a specific row from an ODS file, but when I use fid=3, I don't get anything back.  Less than seems to work correctly, but greater than and equal return incorrect rows or nothing. 

As an example, if I run the following, the first select comes back with all the rows in the file, but trying to select a specific row returns 0 rows.

    create server httpodsserver foreign data wrapper ogr_fdw options (datasource '/vsicurl/https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/561330/reccrime-offence-ref.ods', format 'ODS', config_options 'OGR_ODS_HEADERS=FORCE');
    create schema if not exists imported;
    import foreign schema ogr_all from server httpodsserver into imported;
    select * from imported.reference_table___offence;
    select * from imported.reference_table___offence where fid=3;

**Comment by pramsey** [comment]
Does it work if you change to a database source or a shapefile? Trying to tease apart the source type from the internal mechanics. Fid filtering should work, I'm pretty sure, but every data source has its own take on what a "fid" actually is.

**Comment by sfischaber** [comment]
I tried an ESRI Shapefile and and XLS file and your right, they both work fine.  I also tried an XLSX file, and it has the same problem as the ODS one.

**Comment by pramsey** [comment]
I wonder what it is about fid on spreadsheets that is different...

**Comment by pramsey** [comment]
If you wrap your query in a CTE can you *force* it to work?
```sql
WITH raw AS ( SELECT * FROM mytable ) SELECT * FROM raw WHERE fid = 3;
```

**Comment by sfischaber** [comment]
yes, that works.

**Comment by pramsey** [comment]
If you run [ogr2ogr](http://www.gdal.org/ogr2ogr.html) on your spreadsheet and try using the SQL parameter to filter on fid, does the filtering work there? (Can you see what I'm up to here?)

**Comment by sfischaber** [comment]
I can filter using ogr2ogr with the -sql or -fid options; however, it looks like when I use -sql I have to start fid at 0 to get the first data row, but when I do -fid I have to start it at 2 to get the same row.

    ./ogr2ogr -sql "select * from Open_Data where fid=0" -f "CSV" /tmp/data/sql.csv /vsicurl/https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/584466/prc-transferred-cancelled-crime-march2012-march2015-tabs.ods --config OGR_ODS_HEADERS FORCE
    ./ogr2ogr -fid 2 -f "CSV" /tmp/data/fid.csv /vsicurl/https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/584466/prc-transferred-cancelled-crime-march2012-march2015-tabs.ods --config OGR_ODS_HEADERS FORCE
  
I don't know if that makes any sense or helps any.

**Comment by pramsey** [comment]
I think it just points to this being an issue below the level I have control over, the OGR API. It's an inconsistency in the driver, so nothing I can do at the pgsql-ogr level.

**Comment by sfischaber** [comment]
Thanks for the pointers; yes, it was an [issue](https://trac.osgeo.org/gdal/ticket/6788) within gdal.  It's been fixed and should be part of their 2.1.4 release.
